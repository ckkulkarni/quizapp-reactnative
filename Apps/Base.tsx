import React, {useState, useContext} from 'react';
import {
  Text,
  View,
  Button,
  TextInput,
  StyleSheet,
  ScrollView,
} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import SelectDropdown from 'react-native-select-dropdown';
import {QuizContext} from './QuizContext';
export default function Base() {
  const navigation = useNavigation<NativeStackNavigationProp<any>>();
  const {setName, setMail, setPhone, setSelected, setTracker} =
    useContext(QuizContext);
  const [validMail, setValid] = useState<boolean>(true);
  const [validName, setValidName] = useState<boolean>(false);
  const [validPhone, setValidPhone] = useState<boolean>(false);
  const [submitPressed, setSubmit] = useState<boolean>(false);
  const [validSelections, setSelections] = useState<boolean>(false);
  const [validOption, setOption] = useState<boolean>(false);
  const selections = ['ReactJS', 'React Native', 'JavaScript', 'Java'];
  function handleSubmit() {
    setSubmit(true);
  }
  const handleEmailChange = (text: string) => {
    setMail(text);
    const emailFormat = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i;
    setValid(emailFormat.test(text));
  };
  const handleNameChange = (text: string) => {
    setName(text);
    setValidName(text.trim().length > 0);
  };
  const handlePhoneChange = (text: string) => {
    setPhone(text);
    setValidPhone(text.trim().length === 10);
  };

  const isFormValid = validName && validMail && validPhone;
  const startQuiz = isFormValid && submitPressed;
  const track = [true, false];
  return (
    <ScrollView contentContainerStyle={styles.container}>
      <TextInput
        placeholder="Enter your name"
        onChangeText={handleNameChange}
        style={styles.input}
      />
      {!validMail && (
        <Text style={{color: 'red'}}>Enter a valid Email Address</Text>
      )}
      <TextInput
        placeholder="Enter your Email ID"
        onChangeText={handleEmailChange}
        keyboardType="email-address"
        style={[styles.input, !validMail && styles.invalidInput]}
      />
      <TextInput
        placeholder="Enter your Phone Number"
        keyboardType="numeric"
        onChangeText={handlePhoneChange}
        style={styles.input}
      />
      <Text>Select your preferred language: </Text>
      <SelectDropdown
        data={selections}
        onSelect={(selectedItem, index) => {
          setSelected(selectedItem);
        }}
        searchInputTxtColor={'blue'}
      />
      <Text>Track Score?</Text>
      <SelectDropdown
        data={track}
        onSelect={(tracking, index) => {
          setTracker(tracking);
        }}
        searchInputTxtColor={'blue'}
      />
      <Button title="Submit" onPress={handleSubmit} disabled={!isFormValid} />
      <Button
        disabled={!startQuiz}
        title="Click Here to Start Quiz"
        onPress={() => navigation.navigate('Question 1')}
      />
    </ScrollView>
  );
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  input: {
    borderWidth: 1,
    borderColor: 'gray',
    borderRadius: 5,
    padding: 10,
    width: '80%',
    marginVertical: 10,
  },
  invalidInput: {
    borderColor: 'red',
  },
  selector: {
    backgroundColor: 'blue',
    padding: 10,
  },
});
