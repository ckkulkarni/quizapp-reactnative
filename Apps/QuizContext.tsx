import React from 'react';

export const QuizContext = React.createContext<{
  score: number;
  setScore: React.Dispatch<React.SetStateAction<number>>;
  name: string;
  setName: React.Dispatch<React.SetStateAction<string>>;
  email: string;
  setMail: React.Dispatch<React.SetStateAction<string>>;
  phone: string;
  setPhone: React.Dispatch<React.SetStateAction<string>>;
  selectedValue: string;
  setSelected: React.Dispatch<React.SetStateAction<string>>;
  trackScore: boolean;
  setTracker: React.Dispatch<React.SetStateAction<boolean>>;
}>({
  score: 0,
  setScore: () => {},
  name: '',
  setName: () => {},
  email: '',
  setMail: () => {},
  phone: '',
  setPhone: () => {},
  selectedValue: '',
  setSelected: () => {},
  trackScore: false,
  setTracker: () => {},
});
