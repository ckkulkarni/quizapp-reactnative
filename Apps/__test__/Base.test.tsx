import Base from '../Base';
import Adapter from 'enzyme-adapter-react-16';
import {mount, shallow} from 'enzyme';
import * as enzyme from 'enzyme';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {TextInput, Button} from 'react-native';
import {QuizContext} from './../QuizContext';
enzyme.configure({adapter: new Adapter()});
const navigation = useNavigation<NativeStackNavigationProp<any>>();

jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: jest.fn().mockReturnValue({
      navigate: jest.fn(),
    }),
  };
});
describe('should render a form with three text inputs, two selections, and one submit', () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallow(<Base />);
  });
  it('component renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should update the name state upon text input', () => {
    expect(wrapper.find('TextInput')).toBeTruthy();
    expect(wrapper.find('SelectDropdown')).toBeTruthy();
    expect(wrapper.find('Button')).toBeTruthy();
  });
  it('should disable the submit button when the form  does not have all the inputs', () => {
    expect(wrapper.find('Button').at(0).prop('disabled')).toBe(true);
  });
  it('should enable the submit button when the form has all the inputs', () => {
    wrapper.find('TextInput').at(0).simulate('changeText', 'chandrakanth');
    wrapper.find('TextInput').at(1).simulate('changeText', 'chandra@kanth.com');
    wrapper.find('TextInput').at(2).simulate('changeText', '1234567890');
    expect(wrapper.find('Button').at(0).prop('disabled')).toBe(false);
  });
  it('on quiz start button, should navigate to the next screen, aka, the quiz should begin', () => {
    wrapper.find('Button').at(0).props().onPress();
    wrapper.find(Button).at(1).props().onPress();
    wrapper.update();
    expect(navigation.navigate).toBeCalledWith('Question 1');
  });
});
