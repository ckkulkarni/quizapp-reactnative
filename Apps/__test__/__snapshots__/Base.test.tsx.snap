// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a form with three text inputs, two selections, and one submit component renders correctly 1`] = `ShallowWrapper {}`;
